akka-persistence-query/src/main/scala/akka/persistence/query/typed/internal/EventsBySliceFirehose.scala

```diff
@@ -0,0 +1,773 @@
+            else
+              existing.history.tail :+ offset // drop one, add one
+          val newTracking = existing.copy(history = newHistory)
+          if (consumerTracking.replace(consumerId, existingTracking, newTracking))
+            logUpdateConsumerTracking(consumerId, now, newTracking)
+          else
+            // concurrent update, try again
+            updateConsumerTracking(consumerId, now, offset, consumerKillSwitch)
+      }
+    }
+
+    private def logUpdateConsumerTracking(consumerId: String, now: Instant, tracking: ConsumerTracking): Unit = {
+      if (settings.verboseLogging && log.isDebugEnabled) {
+        val trackingValues = consumerTrackingValues()
+        if (trackingValues.size > 1) {
+          val slowestConsumer = trackingValues.minBy(_.offsetTimestamp)
+          val fastestConsumer = trackingValues.maxBy(_.offsetTimestamp)
+          val behind = elementsBehind(fastestConsumer.history, slowestConsumer.history)
+          if (behind > 0) {
+            val diffSlowestFastestsMillis = fastestConsumer.offsetTimestamp.toEpochMilli - slowestConsumer.offsetTimestamp.toEpochMilli
+            val fastestLagMillis = now.toEpochMilli - fastestConsumer.offsetTimestamp.toEpochMilli
+
+            val diffFastest = fastestConsumer.offsetTimestamp.toEpochMilli - tracking.offsetTimestamp.toEpochMilli
+            val diffFastestStr =
+              if (diffFastest > 0) s"behind fastest [$diffFastest] ms"
+              else if (diffFastest < 0) s"ahead of fastest [$diffFastest] ms" // not possible
+              else "same as fastest"
+            val diffSlowest = slowestConsumer.offsetTimestamp.toEpochMilli - tracking.offsetTimestamp.toEpochMilli
+            val diffSlowestStr =
+              if (diffSlowest > 0) s"behind slowest [$diffSlowest] ms" // not possible
+              else if (diffSlowest < 0) s"ahead of slowest [${-diffSlowest}] ms"
+              else "same as slowest"
+            val consumerBehind = elementsBehind(fastestConsumer.history, tracking.history)
+
+            log.debug(
+              s"Firehose entityType [$entityType] sliceRange [$sliceRangeStr] updateConsumerTracking [$consumerId], " +
+              s"behind [$consumerBehind] events from fastest, " +
+              s"$diffFastestStr, $diffSlowestStr, " +
+              s"slowest [${fastestConsumer.consumerId}] is behind [$behind] " +
+              s"of fastest [${slowestConsumer.consumerId}], [$diffSlowestFastestsMillis] ms. " +
+              s"Consumer lag of fastest [$fastestLagMillis] ms.")
+          }
+        }
+      }
+    }
+
+    def detectSlowConsumers(now: Instant): Unit = {
+      val trackingValues = consumerTrackingValues()
+      if (trackingValues.size > 1) {
+        val slowestConsumer = trackingValues.minBy(_.offsetTimestamp)
+        val fastestConsumer = trackingValues.maxBy(_.offsetTimestamp)
+
+        val behind = elementsBehind(fastestConsumer.history, slowestConsumer.history)
```

Comment:
Hmmm, the slowest consumer will hold back the fastest so that it should always have the timestamp of the slower in its history, because we keep as long offset history as the size of the buffer?

Reply:
Exactly. Not great to have to keep the history for each consumer but I couldn't think of another way. It's just small objects, and probably held in memory by other things.

There is a case when the fastest is a new consumer and then it wouldn't have the full history yet. Then it keeps going.